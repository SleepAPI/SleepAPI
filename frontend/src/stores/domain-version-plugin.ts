import type { PiniaPluginContext } from 'pinia'
import { DOMAIN_VERSION } from 'sleepapi-common'

const domainVersionPlugin = (context: PiniaPluginContext) => {
  const { store } = context

  const cachedVersion = store.domainVersion

  if (cachedVersion && +cachedVersion !== DOMAIN_VERSION) {
    if (typeof store.outdate === 'function') {
      store.outdate()
    }
  }
}

export default domainVersionPlugin
