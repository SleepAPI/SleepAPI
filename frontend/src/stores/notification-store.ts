import { defineStore } from 'pinia'

export interface NotificationState {
  showTeamNameNotification: boolean
}

export const useNotificationStore = defineStore('notification', {
  state: (): NotificationState => {
    return {
      showTeamNameNotification: true
    }
  },
  actions: {
    hideTeamNameNotification() {
      this.showTeamNameNotification = false
    }
  },
  persist: true
})
