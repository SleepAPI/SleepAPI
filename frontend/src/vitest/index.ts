export * from './mocks'
export * from './setup'
