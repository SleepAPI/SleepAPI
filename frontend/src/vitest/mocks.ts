import { ingredient, nature, pokemon, type PokemonInstanceExt } from 'sleepapi-common'

export function createMockPokemon(attrs?: Partial<PokemonInstanceExt>): PokemonInstanceExt {
  return {
    name: 'Bubbles',
    externalId: 'external-id',
    pokemon: pokemon.PIKACHU,
    carrySize: 10,
    ingredients: [{ level: 1, ingredient: ingredient.FANCY_APPLE }],
    level: 10,
    ribbon: 0,
    nature: nature.BASHFUL,
    saved: false,
    shiny: false,
    skillLevel: 1,
    subskills: [],
    version: 1,
    ...attrs
  }
}
