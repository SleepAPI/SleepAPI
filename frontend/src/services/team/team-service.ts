import serverAxios from '@/router/server-axios'
import { PokemonInstanceUtils } from '@/services/utils/pokemon-instance-utils'
import { usePokemonStore } from '@/stores/pokemon/pokemon-store'
import { useTeamStore } from '@/stores/team/team-store'
import {
  MAX_TEAM_MEMBERS,
  type TeamCombinedProduction,
  type TeamInstance,
  type TeamProductionExt
} from '@/types/member/instanced'
import axios from 'axios'
import {
  berry,
  type BerrySet,
  type CalculateTeamResponse,
  type GetTeamsResponse,
  type IngredientSet,
  type PokemonInstanceExt,
  type PokemonInstanceIdentity,
  type TeamSettings,
  type UpsertTeamMemberResponse,
  type UpsertTeamMetaRequest,
  type UpsertTeamMetaResponse
} from 'sleepapi-common'

class TeamServiceImpl {
  public async createOrUpdateTeam(index: number, teamInfo: UpsertTeamMetaRequest) {
    const response = await serverAxios.put<UpsertTeamMetaResponse>(`team/meta/${index}`, teamInfo)

    return response.data
  }

  public async createOrUpdateMember(params: {
    teamIndex: number
    memberIndex: number
    member: PokemonInstanceExt
  }) {
    const { teamIndex, memberIndex, member } = params

    const request = PokemonInstanceUtils.toUpsertTeamMemberRequest(member)

    return await serverAxios.put<UpsertTeamMemberResponse>(
      `team/${teamIndex}/member/${memberIndex}`,
      request
    )
  }

  public async getTeams(): Promise<TeamInstance[]> {
    const response = await serverAxios.get<GetTeamsResponse>('team')

    const serverTeams = response.data.teams
    const teamStore = useTeamStore()
    const pokemonStore = usePokemonStore()

    const teams: TeamInstance[] = []
    for (let teamIndex = 0; teamIndex < teamStore.maxAvailableTeams; teamIndex++) {
      const serverTeam = serverTeams.find((team) => team.index === teamIndex)
      if (!serverTeam) {
        const emptyTeam: TeamInstance = {
          index: teamIndex,
          memberIndex: 0,
          name: `Helper team ${teamIndex + 1}`,
          camp: false,
          bedtime: '21:30',
          wakeup: '06:00',
          recipeType: 'curry',
          favoredBerries: [],
          version: 0,
          members: new Array(MAX_TEAM_MEMBERS).fill(undefined),
          production: undefined
        }
        teams.push(emptyTeam)
      } else {
        const members: (string | undefined)[] = []
        for (let memberIndex = 0; memberIndex < MAX_TEAM_MEMBERS; memberIndex++) {
          const serverMember = serverTeam.members.find(
            (member) => member.memberIndex === memberIndex
          )

          if (!serverMember) {
            members.push(undefined)
          } else {
            const cachedMember = PokemonInstanceUtils.toPokemonInstanceExt(serverMember)
            pokemonStore.upsertLocalPokemon(cachedMember)
            members.push(cachedMember.externalId)
          }
        }

        const favoredBerries: berry.Berry[] = []
        if (serverTeam.favoredBerries?.length === 1 && serverTeam.favoredBerries[0] === 'all') {
          favoredBerries.push(...berry.BERRIES)
        } else {
          for (const berryName of serverTeam.favoredBerries ?? []) {
            const favoredBerry = berry.BERRIES.find((berry) => berry.name === berryName)
            if (!favoredBerry) {
              console.error(
                `Couldn't find favored berry with name: ${berryName}, contact developer`
              )
              continue
            }
            favoredBerries.push(favoredBerry)
          }
        }
        const instancedTeam: TeamInstance = {
          index: serverTeam.index,
          memberIndex: teamStore.teams[serverTeam.index]?.memberIndex ?? 0,
          name: serverTeam.name,
          camp: serverTeam.camp,
          bedtime: serverTeam.bedtime,
          wakeup: serverTeam.wakeup,
          recipeType: serverTeam.recipeType,
          favoredBerries,
          version: serverTeam.version,
          members,
          production: undefined
        }
        teams.push(instancedTeam)
      }
    }

    return teams
  }

  public async deleteTeam(index: number) {
    return await serverAxios.delete(`team/${index}`)
  }

  public async removeMember(params: { teamIndex: number; memberIndex: number }) {
    const { teamIndex, memberIndex } = params
    await serverAxios.delete(`team/${teamIndex}/member/${memberIndex}`)
  }

  // TODO: not tested?
  public async calculateProduction(params: {
    members: PokemonInstanceExt[]
    settings: TeamSettings
  }) {
    const { members, settings } = params
    if (members.length === 0) {
      return undefined
    }

    const parsedMembers: PokemonInstanceIdentity[] = members.map((member) => ({
      pokemon: member.pokemon.name,
      nature: member.nature.name,
      subskills: member.subskills.map((subskill) => ({
        level: subskill.level,
        subskill: subskill.subskill.name
      })),
      ingredients: member.ingredients.map((ingredient) => ({
        level: ingredient.level,
        ingredient: ingredient.ingredient.name
      })),
      carrySize: member.carrySize,
      level: member.level,
      ribbon: member.ribbon,
      skillLevel: member.skillLevel,
      externalId: member.externalId
    }))

    const response = await axios.post<CalculateTeamResponse>('/api/calculator/team', {
      members: parsedMembers,
      settings
    })

    const teamBerries: BerrySet[] = response.data.members.flatMap((member) =>
      member.berries ? member.berries : []
    )
    const teamIngredients: IngredientSet[] = response.data.members.flatMap(
      (member) => member.ingredients
    )
    const teamProduction: TeamCombinedProduction = {
      berries: teamBerries,
      ingredients: teamIngredients,
      cooking: response.data.cooking
    }

    const pokemonStore = usePokemonStore()
    const result: TeamProductionExt = {
      members: response.data.members.map((member) => {
        if (!member.externalId) {
          console.error('Unidentified member in team calculation, contact developer')
          throw new Error('Unidentified member in team calculation')
        }
        return {
          berries: member.berries,
          ingredients: member.ingredients,
          skillAmount: member.skillAmount,
          skillProcs: member.skillProcs,
          member: pokemonStore.getPokemon(member.externalId)
        }
      }),
      team: teamProduction
    }

    return result
  }
}

export const TeamService = new TeamServiceImpl()
