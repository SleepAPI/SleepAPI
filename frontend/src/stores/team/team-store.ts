import { TeamService } from '@/services/team/team-service'
import { randomName } from '@/services/utils/name-utils'
import { usePokemonStore } from '@/stores/pokemon/pokemon-store'
import { useUserStore } from '@/stores/user-store'
import { MAX_TEAMS, MAX_TEAM_MEMBERS, type TeamInstance } from '@/types/member/instanced'
import type { TimeWindowDay } from '@/types/time/time-window'
import { defineStore } from 'pinia'
import {
  DOMAIN_VERSION,
  berry,
  mainskill,
  subskill,
  uuid,
  type PokemonInstanceExt,
  type RecipeType,
  type TeamSettings
} from 'sleepapi-common'

export interface TeamState {
  currentIndex: number
  maxAvailableTeams: number
  loadingTeams: boolean
  loadingMembers: boolean[]
  domainVersion: number
  timeWindow: TimeWindowDay
  tab: 'overview' | 'members' | 'cooking'
  teams: TeamInstance[]
}

export const useTeamStore = defineStore('team', {
  state: (): TeamState => ({
    currentIndex: 0,
    maxAvailableTeams: MAX_TEAMS,
    loadingTeams: false,
    loadingMembers: [false, false, false, false, false],
    domainVersion: 0,
    timeWindow: '24H',
    tab: 'overview',
    teams: [
      {
        index: 0,
        memberIndex: 0,
        name: 'Log in to save your teams',
        camp: false,
        bedtime: '21:30',
        wakeup: '06:00',
        recipeType: 'curry',
        favoredBerries: [],
        version: 0,
        members: new Array(MAX_TEAM_MEMBERS).fill(undefined),
        production: undefined
      }
    ]
  }),
  getters: {
    getCurrentTeam: (state) => state.teams[state.currentIndex],
    getPokemon: (state) => {
      return (memberIndex: number) => {
        const pokemonStore = usePokemonStore()
        const pokemonExternalId = state.teams[state.currentIndex].members[memberIndex]
        return pokemonExternalId != null ? pokemonStore.getPokemon(pokemonExternalId) : undefined
      }
    },
    getTeamSize: (state) => state.teams[state.currentIndex].members.filter(Boolean).length,
    getMemberLoading(state) {
      return (memberIndex: number) => {
        return state.loadingMembers[memberIndex]
      }
    },
    getCurrentMember: (state) => {
      const currentTeam = state.teams[state.currentIndex]
      return currentTeam.production?.members.at(currentTeam.memberIndex)?.memberExternalId
    }
  },
  actions: {
    migrate() {
      if (!this.timeWindow) {
        this.timeWindow = '24H'
      }
      if (!this.tab) {
        this.tab = 'overview'
      }
      if (this.maxAvailableTeams < MAX_TEAMS) {
        this.maxAvailableTeams = MAX_TEAMS
      }

      for (const team of this.teams) {
        if (!team.memberIndex) {
          team.memberIndex = 0
        }
      }
    },
    outdate() {
      for (const team of this.teams) {
        team.production = undefined
      }
      this.domainVersion = DOMAIN_VERSION
    },
    async syncTeams() {
      const userStore = useUserStore()
      const pokemonStore = usePokemonStore()

      if (userStore.loggedIn) {
        // grab previous teams so we may compare it to updated teams from server
        const previousTeams = this.teams.map((team) => ({
          ...team,
          members: team.members.map((member) =>
            member ? pokemonStore.getPokemon(member) : undefined
          )
        }))

        // get updated teams from server
        // this also already updates the pokemon in pokemonStore
        const serverTeams = await TeamService.getTeams()

        // overwrite cached teams with data from server
        this.teams = serverTeams

        for (const serverTeam of serverTeams) {
          const previousTeam = previousTeams.find((team) => team.index === serverTeam.index)
          if (!previousTeam || previousTeam.version !== serverTeam.version) {
            // this team does not exist on this device previously or this team has been updated on other device
            // keep team's production undefined
            continue
          } else {
            // check if any individual members have been updated, if so set updated to true
            let memberUpdated = false
            for (let i = 0; i < serverTeam.members.length; i++) {
              const updatedMember = serverTeam.members[i]
              const populatedUpdateMember = updatedMember
                ? pokemonStore.getPokemon(updatedMember)
                : undefined

              const previousMember = previousTeam.members.find(
                (member, previousIndex) =>
                  previousIndex === i && member?.externalId === populatedUpdateMember?.externalId
              )

              // if the member is not the same uuid or the member is the same, but version mismatches we also re-sim
              if (previousMember && populatedUpdateMember?.version !== previousMember.version) {
                memberUpdated = true
                break
              }
            }

            // if neither team nor members have been update we copy production from cache
            if (!memberUpdated) {
              this.teams[serverTeam.index].production = previousTeam.production
            }
          }
        }
      }
    },
    next() {
      this.currentIndex = (this.currentIndex + 1) % this.teams.length
    },
    prev() {
      this.currentIndex = (this.currentIndex - 1 + this.teams.length) % this.teams.length
    },
    async updateTeam() {
      const userStore = useUserStore()
      if (userStore.loggedIn) {
        try {
          const {
            favoredBerries: teamBerries,
            name,
            camp,
            bedtime,
            wakeup,
            recipeType
          } = this.getCurrentTeam

          const anyFavoredBerries = teamBerries?.length > 0
          const favoredBerries = anyFavoredBerries
            ? teamBerries.length === berry.BERRIES.length
              ? ['all']
              : teamBerries.map((b) => b.name)
            : undefined

          const { version } = await TeamService.createOrUpdateTeam(this.currentIndex, {
            name,
            camp,
            bedtime,
            wakeup,
            recipeType,
            favoredBerries
          })

          this.getCurrentTeam.version = version
        } catch {
          console.error('Error updating teams')
        }
      }
    },
    async deleteTeam() {
      const userStore = useUserStore()

      const newName = userStore.loggedIn
        ? `Helper team ${this.currentIndex + 1}`
        : 'Log in to save your teams'

      this.teams[this.currentIndex] = {
        index: this.currentIndex,
        memberIndex: 0,
        camp: false,
        name: newName,
        bedtime: '21:30',
        wakeup: '06:00',
        recipeType: 'curry',
        favoredBerries: [],
        version: 0,
        members: new Array(MAX_TEAM_MEMBERS).fill(undefined),
        production: undefined
      }

      if (userStore.loggedIn) {
        try {
          await TeamService.deleteTeam(this.currentIndex)
        } catch {
          console.error('Error updating teams')
        }
      }
    },
    async updateTeamMember(updatedMember: PokemonInstanceExt, memberIndex: number) {
      this.loadingMembers[memberIndex] = true

      const userStore = useUserStore()
      const pokemonStore = usePokemonStore()

      pokemonStore.upsertLocalPokemon(updatedMember)

      if (userStore.loggedIn) {
        try {
          TeamService.createOrUpdateMember({
            teamIndex: this.currentIndex,
            memberIndex,
            member: updatedMember
          })
        } catch (error) {
          console.error('Error updating teams')
        }
      }

      this.teams[this.currentIndex].members[memberIndex] = updatedMember.externalId

      await this.calculateProduction(this.currentIndex)
      // reset single production to trigger radar chart recalc
      if (this.isSupportMember(updatedMember)) {
        this.resetCurrentTeamSingleProduction()
      } else if (
        this.teams[this.currentIndex]?.production &&
        this.teams[this.currentIndex].production?.members[memberIndex]
      ) {
        this.teams[this.currentIndex].production!.members[memberIndex].singleProduction = undefined
      }

      this.loadingMembers[memberIndex] = false
    },
    async calculateProduction(teamIndex: number) {
      const pokemonStore = usePokemonStore()
      this.loadingTeams = true

      const members: PokemonInstanceExt[] = []
      for (const member of this.teams[teamIndex].members) {
        if (member) {
          const pokemon = pokemonStore.getPokemon(member)
          pokemon && members.push(pokemon)
        }
      }
      const settings: TeamSettings = {
        camp: this.teams[teamIndex].camp,
        bedtime: this.teams[teamIndex].bedtime,
        wakeup: this.teams[teamIndex].wakeup
      }
      try {
        const newProduction = await TeamService.calculateProduction({ members, settings })

        const existingProduction = this.teams[teamIndex].production

        if (!existingProduction || !newProduction) {
          this.teams[teamIndex].production = newProduction
        } else {
          // preserve the cached single production results
          const mergedMembers = newProduction.members.map((newMemberProduction, index) => {
            const existingMemberProduction = existingProduction.members[index]

            if (existingMemberProduction && existingMemberProduction.singleProduction) {
              return {
                ...newMemberProduction,
                singleProduction: existingMemberProduction.singleProduction
              }
            }

            return newMemberProduction
          })

          this.teams[teamIndex].production = {
            ...newProduction,
            members: mergedMembers
          }
        }
      } catch {
        console.error('Could not calculate production, contact developer')
      }
      this.loadingTeams = false
    },
    async duplicateMember(memberIndex: number) {
      const existingMember = this.getPokemon(memberIndex)
      const currentTeam = this.getCurrentTeam

      const openSlotIndex = currentTeam.members.findIndex((member) => member == null) // checks for both null and undefined
      if (openSlotIndex === -1 || existingMember == null) {
        console.error("No open slot or member can't be found")
        return
      }
      this.loadingMembers[openSlotIndex] = true

      const duplicatedMember: PokemonInstanceExt = {
        ...existingMember,
        version: 0,
        saved: false,
        externalId: uuid.v4(),
        name: randomName(12, existingMember.gender)
      }
      await this.updateTeamMember(duplicatedMember, openSlotIndex)
      this.loadingMembers[openSlotIndex] = false
    },
    async removeMember(memberIndex: number) {
      this.loadingMembers[memberIndex] = true

      const userStore = useUserStore()
      const pokemonStore = usePokemonStore()

      // members array will be reduced by one
      this.getCurrentTeam.memberIndex = Math.max(this.getCurrentTeam.memberIndex - 1, 0)

      if (userStore.loggedIn) {
        try {
          TeamService.removeMember({
            teamIndex: this.currentIndex,
            memberIndex
          })
        } catch (error) {
          console.error('Error updating teams')
        }
      }

      const member = this.getPokemon(memberIndex)
      if (member != null) {
        if (this.isSupportMember(member)) {
          this.resetCurrentTeamSingleProduction()
        }
        pokemonStore.removePokemon(member.externalId, 'team')
      }

      this.teams[this.currentIndex].members[memberIndex] = undefined

      this.loadingMembers[memberIndex] = false
      await this.calculateProduction(this.currentIndex)
    },
    async toggleCamp() {
      this.getCurrentTeam.camp = !this.getCurrentTeam.camp

      this.resetCurrentTeamSingleProduction()
      this.updateTeam()
      await this.calculateProduction(this.currentIndex)
    },
    async updateSleep(params: { bedtime: string; wakeup: string }) {
      const { bedtime, wakeup } = params
      this.getCurrentTeam.bedtime = bedtime
      this.getCurrentTeam.wakeup = wakeup

      this.resetCurrentTeamSingleProduction()
      this.updateTeam()
      await this.calculateProduction(this.currentIndex)
    },
    async updateRecipeType(recipeType: RecipeType) {
      this.getCurrentTeam.recipeType = recipeType

      this.updateTeam()
    },
    async updateFavoredBerries(berries: berry.Berry[]) {
      this.getCurrentTeam.favoredBerries = berries

      this.updateTeam()
    },
    resetCurrentTeamSingleProduction() {
      if (this.getCurrentTeam.production) {
        this.getCurrentTeam.production.members.forEach((member) => {
          if (member.singleProduction) {
            member.singleProduction = undefined
          }
        })
      }
    },
    isSupportMember(member: PokemonInstanceExt) {
      const hbOrErb = member.subskills.some(
        (s) =>
          (s.subskill.name.toLowerCase() === subskill.ENERGY_RECOVERY_BONUS.name.toLowerCase() ||
            s.subskill.name.toLowerCase() === subskill.HELPING_BONUS.name.toLowerCase()) &&
          s.level <= member.level
      )
      const supportSkill = [
        mainskill.ENERGIZING_CHEER_S,
        mainskill.ENERGY_FOR_EVERYONE,
        mainskill.HELPER_BOOST,
        mainskill.EXTRA_HELPFUL_S,
        mainskill.METRONOME
      ].some((s) => s.name.toLowerCase() === member.pokemon.skill.name.toLowerCase())

      return hbOrErb || supportSkill
    }
  },
  persist: true
})
