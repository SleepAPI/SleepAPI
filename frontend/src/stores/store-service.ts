import { useComparisonStore } from '@/stores/comparison-store/comparison-store'
import { useNotificationStore } from '@/stores/notification-store'
import { usePokedexStore } from '@/stores/pokedex-store/pokedex-store'
import { usePokemonStore } from '@/stores/pokemon/pokemon-store'
import { useTeamStore } from '@/stores/team/team-store'
import { useUserStore } from '@/stores/user-store'

export function clearCacheAndLogout() {
  localStorage.clear()

  const userStore = useUserStore()
  const teamStore = useTeamStore()
  const pokemonStore = usePokemonStore()
  const pokedexStore = usePokedexStore()
  const notificationStore = useNotificationStore()
  const comparisonStore = useComparisonStore()

  userStore.$reset()
  teamStore.$reset()
  pokemonStore.$reset()
  pokedexStore.$reset()
  notificationStore.$reset()
  comparisonStore.$reset()
}

export function clearCacheKeepLogin() {
  const teamStore = useTeamStore()
  const pokemonStore = usePokemonStore()
  const pokedexStore = usePokedexStore()
  const notificationStore = useNotificationStore()
  const comparisonStore = useComparisonStore()

  teamStore.$reset()
  pokemonStore.$reset()
  pokedexStore.$reset()
  notificationStore.$reset()
  comparisonStore.$reset()
}

export function migrateStores() {
  const teamStore = useTeamStore()
  const comparisonStore = useComparisonStore()
  teamStore.migrate()
  comparisonStore.migrate()
}
