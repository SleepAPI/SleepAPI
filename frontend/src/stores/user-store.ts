import router from '@/router/router'
import { GoogleService } from '@/services/login/google-service'
import { clearCacheAndLogout } from '@/stores/store-service'
import { defineStore } from 'pinia'
import { MAX_ISLAND_BONUS, type LoginResponse } from 'sleepapi-common'
import { googleLogout } from 'vue3-google-login'

export interface UserState {
  name: string
  avatar: string | null
  email: string | null
  tokens: TokenInfo | null
  externalId: string | null
}

export interface TokenInfo {
  expiryDate: number
  accessToken: string
  refreshToken: string
}

export const useUserStore = defineStore('user', {
  state: (): UserState => {
    return {
      name: 'Guest',
      avatar: null,
      email: null,
      tokens: null,
      externalId: null
    }
  },
  getters: {
    // loggedIn doesnt check if access token valid/expired, just that the user is logged in
    loggedIn: (state) => state.tokens !== null,
    islandBonus: (state) => 1 + MAX_ISLAND_BONUS / 100 // TODO: user should probably hold island bonuses and recipe levels
  },
  actions: {
    setUserData(userData: { name: string; avatar?: string; email: string; externalId: string }) {
      this.name = userData.name
      this.avatar = userData.avatar ?? 'default'
      this.email = userData.email
      this.externalId = userData.externalId
    },
    setTokens(tokens: TokenInfo) {
      this.tokens = tokens
    },
    async login(authCode: string) {
      const loginResponse: LoginResponse = await GoogleService.login(authCode)
      this.setTokens({
        accessToken: loginResponse.access_token,
        refreshToken: loginResponse.refresh_token,
        expiryDate: loginResponse.expiry_date
      })
      this.setUserData({
        name: loginResponse.name,
        avatar: loginResponse.avatar,
        email: loginResponse.email,
        externalId: loginResponse.externalId
      })

      // Refresh the current page
      router.go(0)
    },
    async refresh() {
      try {
        const tokens = this.tokens
        if (tokens?.expiryDate) {
          if (Date.now() > tokens.expiryDate) {
            const { refreshToken } = tokens
            const { access_token, expiry_date } = await GoogleService.refresh(refreshToken)
            this.setTokens({
              accessToken: access_token,
              refreshToken,
              expiryDate: expiry_date
            })
          }
        } else {
          this.logout()
        }
      } catch {
        this.logout()
      }
    },
    logout() {
      clearCacheAndLogout()
      googleLogout()
      router.push('/')
    }
  },
  persist: true
})
