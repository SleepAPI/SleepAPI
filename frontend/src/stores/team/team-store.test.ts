import { TeamService } from '@/services/team/team-service'
import { useTeamStore } from '@/stores/team/team-store'
import { useUserStore } from '@/stores/user-store'
import { createPinia, setActivePinia } from 'pinia'
import { beforeEach, describe, expect, it, vi } from 'vitest'

vi.mock('@/services/team/team-service', () => ({
  TeamService: {
    getTeams: vi.fn(),
    createOrUpdateTeam: vi.fn()
  }
}))

describe('Team Store', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
  })

  it('should have expected default state', () => {
    const teamStore = useTeamStore()
    expect(teamStore.$state).toMatchInlineSnapshot(`
      {
        "currentIndex": 0,
        "loadingTeams": true,
        "maxAvailableTeams": 5,
        "teams": [
          {
            "camp": false,
            "index": 0,
            "name": "Log in to save your teams",
          },
        ],
      }
    `)
  })

  it('should populate teams correctly when user is logged in', async () => {
    const teamStore = useTeamStore()
    const userStore = useUserStore()
    userStore.setTokens({
      accessToken: 'token1',
      expiryDate: 10,
      refreshToken: 'token2'
    })
    TeamService.getTeams = vi.fn().mockResolvedValue([
      { index: 0, name: 'Team 1', camp: true },
      { index: 1, name: 'Team 2', camp: false }
    ])

    await teamStore.populateTeams()

    expect(TeamService.getTeams).toHaveBeenCalled()
    expect(teamStore.$state.teams).toMatchInlineSnapshot(`
      [
        {
          "camp": true,
          "index": 0,
          "name": "Team 1",
        },
        {
          "camp": false,
          "index": 1,
          "name": "Team 2",
        },
        {
          "camp": false,
          "index": 2,
          "name": "Helper team 3",
        },
        {
          "camp": false,
          "index": 3,
          "name": "Helper team 4",
        },
        {
          "camp": false,
          "index": 4,
          "name": "Helper team 5",
        },
      ]
    `)
  })

  it('should set loadingTeams to false if user is not logged in', async () => {
    const teamStore = useTeamStore()
    const userStore = useUserStore()
    userStore.tokens = null

    await teamStore.populateTeams()

    expect(teamStore.loadingTeams).toBeFalsy()
    expect(teamStore.$state.teams[0].name).toBe('Log in to save your teams')
  })

  it('should update current team name correctly', async () => {
    const teamStore = useTeamStore()
    const userStore = useUserStore()
    userStore.setTokens({
      accessToken: 'token1',
      expiryDate: 10,
      refreshToken: 'token2'
    })
    teamStore.teams = [{ index: 0, name: 'Old Team Name', camp: false }]

    await teamStore.updateTeamName('New Team Name')

    expect(teamStore.teams[0].name).toBe('New Team Name')
    expect(teamStore.getCurrentTeam.name).toBe('New Team Name')
    expect(TeamService.createOrUpdateTeam).toHaveBeenCalledWith(0, {
      name: 'New Team Name',
      camp: false
    })
  })

  it('should reset the state correctly', () => {
    const teamStore = useTeamStore()
    teamStore.teams = [
      { index: 0, name: 'Team 1', camp: false },
      { index: 1, name: 'Team 2', camp: false }
    ]
    teamStore.currentIndex = 1
    teamStore.maxAvailableTeams = 4
    teamStore.loadingTeams = false

    teamStore.reset()

    expect(teamStore.$state).toMatchInlineSnapshot(`
      {
        "currentIndex": 0,
        "loadingTeams": true,
        "maxAvailableTeams": 5,
        "teams": [
          {
            "camp": false,
            "index": 0,
            "name": "Log in to save your teams",
          },
        ],
      }
    `)
  })

  it('should increment currentIndex correctly on next()', () => {
    const teamStore = useTeamStore()
    teamStore.teams = [
      { index: 0, name: 'Team 1', camp: false },
      { index: 1, name: 'Team 2', camp: false }
    ]

    teamStore.next()
    expect(teamStore.currentIndex).toBe(1)

    teamStore.next()
    expect(teamStore.currentIndex).toBe(0)
  })

  it('should decrement currentIndex correctly on prev()', () => {
    const teamStore = useTeamStore()
    teamStore.teams = [
      { index: 0, name: 'Team 1', camp: false },
      { index: 1, name: 'Team 2', camp: false }
    ]

    teamStore.prev()
    expect(teamStore.currentIndex).toBe(1)

    teamStore.prev()
    expect(teamStore.currentIndex).toBe(0)
  })
})
