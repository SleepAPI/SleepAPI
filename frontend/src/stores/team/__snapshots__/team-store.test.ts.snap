// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migrate > shall migrate old teams to new state 1`] = `
[
  {
    "bedtime": "21:10",
    "camp": false,
    "favoredBerries": [],
    "index": 0,
    "memberIndex": 0,
    "memberIvs": {},
    "members": [
      undefined,
      "member1",
      undefined,
      "member2",
      undefined,
    ],
    "name": "Team 1",
    "production": undefined,
    "recipeType": "curry",
    "version": 1,
    "wakeup": "06:00",
  },
  {
    "bedtime": "21:20",
    "camp": false,
    "favoredBerries": [],
    "index": 0,
    "memberIndex": 0,
    "memberIvs": {},
    "members": [
      undefined,
      "member3",
      undefined,
      "member4",
      undefined,
    ],
    "name": "Team 2",
    "production": undefined,
    "recipeType": "curry",
    "version": 1,
    "wakeup": "06:00",
  },
]
`;
